src/execution/execute-afm.ts

```diff
@@ -1,137 +1,251 @@
+                return this.getWholeExecutionResult(executionResponse.links.executionResult)
+                    .then((executionResult: Execution.IExecutionResult | null) => {
+                        return { executionResponse, executionResult };
                     });
             });
     }
 
-    private fetchExecutionResult(pollingUri: string, offset: number[]) {
-        const [uriPart, queryPart] = pollingUri.split(/\?(.+)/);
-        const query = {
-            ...qs.parse(queryPart),
-            limit: getLimit(offset).join(','),
-            offset: offset.join(',')
-        };
-        const finalPollingUri = uriPart + qs.stringify(query, { addQueryPrefix: true });
-        return this.xhr.ajax(finalPollingUri, { method: 'GET' }).then((r) => {
-            if (r.response.status === 204) {
-                return null;
-            }
-            return r.getData();
-        });
+    /**
+     * Get Response from Execution
+     *
+     * @method getExecutionResponse
+     * @param {string} projectId - GD project identifier
+     * @param {AFM.IExecution} execution - See https://github.com/gooddata/gooddata-typings/blob/master/index.ts#L4
+     *
+     * @returns {Promise<Execution.IExecutionResponse>} Promise with `executionResponse`
+     *  See https://github.com/gooddata/gooddata-typings/blob/master/src/Execution.ts#L69
+     */
+    private getExecutionResponse(projectId: string, execution: AFM.IExecution)
+        : Promise<Execution.IExecutionResponse> {
+        validateNumOfDimensions(get(execution, 'execution.resultSpec.dimensions').length);
+        return this.xhr.post(`/gdc/app/projects/${projectId}/executeAfm`, { body: JSON.stringify(execution) })
+            .then(apiResponse => apiResponse.getData())
+            .then(unwrapExecutionResponse);
     }
 
-    private getOnePage(
-        pollingUri: string, offset: number[], prevResultWrapper?: Execution.IExecutionResultWrapper
-    ): any {
-        return this.fetchExecutionResult(pollingUri, offset)
+    /**
+     * Get whole (all pages) Result from Execution
+     *
+     * @method getWholeExecutionResult
+     * @param {string} executionResultUri
+     *
+     * @returns {Promise<Execution.IExecutionResult | null>}
+     *  Promise with `executionResult` or `null` (null means empty response - HTTP 204)
+     *  See https://github.com/gooddata/gooddata-typings/blob/master/src/Execution.ts#L88
+     */
+    private getWholeExecutionResult(executionResultUri: string)
+        : Promise<Execution.IExecutionResult | null> {
+        const numOfDimensions = Number(qs.parse(executionResultUri).dimensions);
+        validateNumOfDimensions(numOfDimensions);
+
+        const limit = Array(numOfDimensions).fill(DEFAULT_LIMIT);
+        const offset = Array(numOfDimensions).fill(0);
+
+        return this.getPage(true, executionResultUri, limit, offset);
+    }
+
+    private getPage(
+        recursive: boolean,
```

Comment:
As I said before, I strongly advise against using same function recursively and not. It can be named `getAllPagesRecursively` which warns every developer to look for recursion.

Nevertheless, it seems the function is not used in not-recursive manner, so there shouldnt be need?

Reply:
@zbycz I guess its preparation for real paging implementation where it's not running recursively all the time

Reply:
Removed non-recursive approach